pkg/objstore/client/factory.go

```diff
@@ -3,39 +3,59 @@ package client
 import (
 	"context"
 	"fmt"
-	"runtime"
 
-	"cloud.google.com/go/storage"
 	"github.com/go-kit/kit/log"
 	"github.com/improbable-eng/thanos/pkg/objstore"
 	"github.com/improbable-eng/thanos/pkg/objstore/gcs"
 	"github.com/improbable-eng/thanos/pkg/objstore/s3"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/common/version"
-	"google.golang.org/api/option"
+	yaml "gopkg.in/yaml.v2"
 )
 
-var ErrNotFound = errors.New("no valid GCS or S3 configuration supplied")
+type objProvider string
+
+const (
+	GCS objProvider = "GCS"
+	S3  objProvider = "S3"
+)
+
+type BucketConfig struct {
+	Type   objProvider `yaml:"type"`
+	Config interface{} `yaml:"config"`
+}
+
+var ErrNotFound = errors.New("not found bucket")
 
 // NewBucket initializes and returns new object storage clients.
-func NewBucket(logger log.Logger, gcsBucket *string, s3Config s3.Config, reg *prometheus.Registry, component string) (objstore.Bucket, error) {
-	if *gcsBucket != "" {
-		gcsOptions := option.WithUserAgent(fmt.Sprintf("thanos-%s/%s (%s)", component, version.Version, runtime.Version()))
-		gcsClient, err := storage.NewClient(context.Background(), gcsOptions)
-		if err != nil {
-			return nil, errors.Wrap(err, "create GCS client")
-		}
-		return objstore.BucketWithMetrics(*gcsBucket, gcs.NewBucket(*gcsBucket, gcsClient, reg), reg), nil
+func NewBucket(logger log.Logger, conf string, reg *prometheus.Registry, component string) (objstore.Bucket, error) {
+	var err error
+	var bucketConf BucketConfig
+	if conf == "" {
+		return nil, ErrNotFound
+	}
+	if err := yaml.Unmarshal([]byte(conf), &bucketConf); err != nil {
+		return nil, errors.Wrap(err, "unmarshal objstore.config")
 	}
 
-	if s3Config.Validate() == nil {
-		b, err := s3.NewBucket(logger, &s3Config, reg, component)
-		if err != nil {
-			return nil, errors.Wrap(err, "create s3 client")
-		}
-		return objstore.BucketWithMetrics(s3Config.Bucket, b, reg), nil
+	config, err := yaml.Marshal(bucketConf.Config)
+	if err != nil {
+		return nil, err
 	}
 
-	return nil, ErrNotFound
+	var bucket objstore.Bucket
+	switch bucketConf.Type {
+	case GCS:
+		bucket, err = gcs.NewBucket(logger, context.Background(), config, reg, component)
+	case S3:
+		bucket, err = s3.NewBucket(logger, config, reg, component)
+	case "":
```

Comment:
This one is not required. It's clearly an error if someone puts config  with `type: ""` (:

Reply:
Done